efficient I2C readout

patman15 opened the pull request:
Hi!

I moved the rgb scaling function to the actual getter function in order to avoid unnecessary calculations in case just raw data is used. Further, I fixed a wrong type in the documentation.

patman15 committed:
fixed struct type in readme

patman15 committed:
optimized raw data readout

moved rgb scaling to actual readout to speed up raw data readout

patman15 committed:
updated version

patman15 committed:
optimized I2C bus access

The sensor readout speed is increased by one larger I2C access and
direct mapping to the internal structure. Function was made public so
that the function can also be used to directly read to user structures.

hideakitai commented on TCS34725.h:
```diff
@@ -60,7 +60,7 @@ class TCS34725_
 
     struct Color { float r, g, b; };
-    struct RawData { uint16_t r, g, b, c; };
+    struct __attribute__ ((packed)) RawData { uint16_t c, r, g, b; } ;
```
I prefer using `union` instead of `struct` in this case

patman15 replied:
I agree here because I learned something. :-)

hideakitai commented on TCS34725.h:
```diff
@@ -133,7 +133,7 @@ class TCS34725_
         if (b)
         {
-            update();
+            readCRGB(&raw_data);
```
No need to change signature

patman15 replied:
I think I can finally update my pull request after clarifying this issue with you. The reason for me to change this was to be able to use a fully IRQ driven program flow. `update` reads a register to see if data is available, which I know it is when I use the external IRQ of the chip. Thus, I made `readCRGB` public (to avoid I2C access of `read8(Reg::STATUS)`).
To give you an idea in my program I use these calls
`  if (dat_avail) {            
    tcs.readCRGB(&data);
    tcs.clearInterrupt();
    dat_avail = false; ...`
and an IRQ handler that sets `dat_avail = true;`
Do you still want to keep it as it is?

patman15 replied:
Now I also remember why I added the pointer as a parameter. Otherwise the calling program always requires a copy operation of the data, i.e. `TCS34725::RawData raw = tcs.raw();` which is unnecessary. Just a thought ...

hideakitai replied:
Sorry I've missed the change about making `update` (`readCRGB`) public. I think it's better to make apis better for  interrupt driven way, but it should be separated from this pull request. I think the main purpose of this pull request is reading data in one i2c request.

About passing the pointer, I think it's not necessary. But I agree with calculating rgb data only when you call `color()`.

hideakitai commented on TCS34725.h:
```diff
@@ -206,16 +206,20 @@ class TCS34725_
+          *((uint8_t *) crgb+i) = wire->read();
+    }    
+
+private:
```
If you use `union` you can change here like:

```C++
    void update()
    {
        wire->beginTransmission(I2C_ADDR);
        wire->write(COMMAND_BIT | (uint8_t) Reg::CDATAL);
        wire->endTransmission();
        wire->requestFrom(I2C_ADDR, sizeof(RawData));
        for (uint8_t i = 0; i < sizeof(RawData); i++)
            raw_data.raw[i] = wire->read();

        if (raw_data.c == 0) clr.r = clr.g = clr.b = 0;
        else
        {
            clr.r = pow((float)raw_data.r / (float)raw_data.c, scaling) * 255.f;
            clr.g = pow((float)raw_data.g / (float)raw_data.c, scaling) * 255.f;
            clr.b = pow((float)raw_data.b / (float)raw_data.c, scaling) * 255.f;
            if (clr.r > 255.f) clr.r = 255.f;
            if (clr.g > 255.f) clr.g = 255.f;
            if (clr.b > 255.f) clr.b = 255.f;
        }
    }
```

patman15 replied:
I agree for the readout, still I would propose to move the color calculation to the function `const Color& color()` itself. If you just want raw data, this calculation just uses unnecessary processor cycles without benefit and it is a transparent change without interface adaptation.

hideakitai commented on TCS34725.h:
```diff
@@ -257,14 +261,14 @@ class TCS34725_
     WireType* wire;
     float scaling {2.5f};
-
+    
```
No need to change here

hideakitai commented on TCS34725.h:
```diff
@@ -257,14 +261,14 @@ class TCS34725_
     bool b_ct_lux_calc {true};
     float lx;
     float color_temp;
+        
```
No need to change here

hideakitai commented on example/example.ino:
```diff
@@ -7,6 +7,7 @@ void setup(void)
     Serial.begin(115200);
 
     Wire.begin();
+    // Wire.setClock(400000); // set maximum sensor readout speed (400kHz)    
```
No need to change here

hideakitai commented on README.md:
```diff
@@ -23,6 +23,7 @@ void setup(void)
     Serial.begin(115200);
 
     Wire.begin();
+    // Wire.setClock(400000); // set maximum sensor readout speed (400kHz)
```
No need to change here

patman15 replied:
My idea here was to give people a hint that you can operate the sensor much faster than with the default. It took me a while to figure that out so I thought a hint would help others.

hideakitai replied:
Thanks, I think it's useful but it's not related to the main purpose of this pull request. So it's better to separate the pull request!

hideakitai commented on README.md:
````diff
@@ -106,7 +107,7 @@ Please see [this library](https://github.com/adafruit/Adafruit_CircuitPython_TCS
 ``` C++
     struct Color { float r, g, b; };
-    struct RawData { uint16_t r, g, b, c; };
+    struct RawData { uint16_t c, r, g, b};
````
I prefer using `union`

hideakitai commented on README.md:
```diff
@@ -132,6 +133,7 @@ Please see [this library](https://github.com/adafruit/Adafruit_CircuitPython_TCS
     void write8(Reg reg, uint8_t value)
     uint8_t read8(Reg reg)
     uint16_t read16(Reg reg)
+    void readCRGB(RawData *crgb)
```
No need to change signature

hideakitai commented on library.json:
```diff
@@ -13,7 +13,7 @@
         "maintainer": true
     },
-    "version": "0.1.1",
+    "version": "0.2.0",
```
I prefer 0.1.1 because there is no major api change

hideakitai commented on library.properties:
```diff
@@ -1,5 +1,5 @@
 name=TCS34725
-version=0.1.1
+version=0.2.0
```
I prefer 0.1.1 because there is no major api change

hideakitai commented:
Good point, but I prefer using `union` instead of `struct`. Please see the comments above and sample I made in another branch https://github.com/hideakitai/TCS34725/blob/optimize_raw_data_read/TCS34725.h

patman15 commented:
Thanks for the feedback, I'll do my best to adapt. Knowing my available time, it usually takes a while ...

hideakitai commented:
@patman15 If you don't have the problem, I will merge my reference branch to master. If you don't have the time to change it, please let me know.

hideakitai commented:
@patman15 Thank you for the response, I think it's much better to separate this pull request into three small pull requests.

- Efficient I2C readout (I2C read in one call, calculate RGB in `color()`)
- Interrupt friendly API improvement
- Add hints to example (Wire.setClock)

And also you don't need to change the version, I will do it later. Thank you for the contribution!

patman15 commented:
@hideakitai ok, thanks for the feedback, I'll split the things and start with the readout and the calculation separation. I'll do then two separate requests for the other things. I appreciate your explanations very much.

patman15 committed:
adapted to review comments

change limited to efficient I2C readout (I2C read in one call, calculate
RGB in color())

patman15 committed:
adapted to review comments

aligned README, example, and properties

patman15 commented:
@hideakitai so, I tried to fix everything to what we agreed. I hope this still fits this PR or should I close and open a completely new one?

hideakitai commented:
I think it's better to open new one or rebase the commits for the cleaner commit logs :)
